### PR TITLE
Add CollectionView support

### PR DESCRIPTION
**Pull request by sergdort**
### Description of change

1. Add `CollectionViewDataSource` which acts as a `dataSource` for the `UICollectionView`
2. Add `CollectionViewSectionDiff ` which uses `FlexibleDiff` to apply diff to the `UICollectionView `
3. Add of usage `Bento` with a `UICollectionView`
3. Add `CollectionViewSupplementaryView` and `CollectionViewCell` as a container for renderable views

### Screenshots

collection_view

### CODE REVIEW

File: Bento/Adapters/CollectionViewDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        self.sections = sections
+        diff.apply(to: collectionView, completion: completion)
+    }
+
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return sections.count
+    }
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return sections[section].rows.count
```
Comment (dmcrodrigues) [inline Bento/Adapters/CollectionViewDataSource.swift]: I think we should rename `rows` to `items` to be more agnostic of any visual representation.

File: Bento/Diff/CollectionViewSectionDiff.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                                      areMetadataEqual: Section.hasEqualMetadata,
+                                      items: { $0.rows },
+                                      itemIdentifier: { $0.id },
+                                      areItemsEqual: ==)
+        apply(diff: diff, to: collectionView, completion: completion)
+    }
+
+    private func apply(diff: SectionedChangeset, to collectionView: UICollectionView, completion: (() -> Void)?) {
+        collectionView.performBatchUpdates({
+            for (item, section) in diff.sections.mutations.enumerated() {
```
Comment (dmcrodrigues) [inline Bento/Diff/CollectionViewSectionDiff.swift]: Maybe we should define this block as a closure and then call `collectionView.performBatchUpdates` at the end with it, I think that would improve a bit the readability.

File: Bento/Diff/CollectionViewSectionDiff.swift [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                if let footerView = collectionView.supplementaryView(forElementKind: UICollectionElementKindSectionFooter.description,
+                                                                     at: IndexPath(row: item, section: section)),
+                    let node = newSections[section].footer {
+                    update(view: footerView, with: node)
+                }
+            }
+
+            collectionView.insertSections(diff.sections.inserts)
+            collectionView.deleteSections(diff.sections.removals)
+            apply(sectionMutations: diff.mutatedSections, to: collectionView)
```
Comment (dmcrodrigues) [inline Bento/Diff/CollectionViewSectionDiff.swift]: [Nit] Maybe we can add this method as an extension of `UICollectionView`?

File: Bento/Views/CollectionViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import UIKit
+
+final class CollectionViewCell: UICollectionViewCell {
```
Comment (dmcrodrigues) [inline Bento/Views/CollectionViewCell.swift]: Call we rename it to `CollectionViewContainerCell`? Just to make it more clear that's not an usual collection view cell. And consequently extend that to `TableViewCell`.

File: Bento/Views/CollectionViewSupplementaryView.swift [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import UIKit
+
+final class CollectionViewSupplementaryView: UICollectionReusableView {
```
Comment (daniloaliberti) [inline Bento/Views/CollectionViewSupplementaryView.swift]: much better! 👌

File: Example/CollectionViewExample/IntroViewController.swift [hunk 6]
```diff
@@ -1,70 +1,82 @@
-import UIKit
+import AVFoundation
+import AVKit
```
Comment (daniloaliberti) [inline Example/CollectionViewExample/IntroViewController.swift]: Any reason for these imports? 
```swift
import AVFoundation
import AVKit
```

File: Bento/Diff/CollectionViewSectionDiff.swift [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    private func apply(sectionMutations: [SectionedChangeset.MutatedSection],
+                       to collectionView: UICollectionView) {
+        for sectionMutation in sectionMutations {
+
+            collectionView.deleteItems(at: sectionMutation.deletedIndexPaths)
+            collectionView.insertItems(at: sectionMutation.insertedIndexPaths)
+            collectionView.perform(moves: sectionMutation.movedCollectionIndexPaths)
+            [sectionMutation.changeset.moves.lazy
```
Comment (mluisbrown) [inline Bento/Diff/CollectionViewSectionDiff.swift]: I found this bit of code really hard to read. Maybe create an intermediate variable `sectionChanges` with everything up to `joined` and then call `sectionChanges.forEach {}` ?

File: Bento/Adapters/CollectionViewDataSource.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    private var sections: [Section<SectionId, ItemId>] = []
+    private weak var collectionView: UICollectionView?
+
+    init(with collectionView: UICollectionView) {
+        sections = []
+        self.collectionView = collectionView
+        super.init()
+        collectionView.dataSource = self
+        /*
+         Force reset of the collection view state to prevent inconsistency on first reload
```
[the remark this answers is not in the document]
Comment (sergdort) [reply Bento/Adapters/CollectionViewDataSource.swift]: Why?

[the remark this answers is not in the document]
Comment (mluisbrown) [reply Bento/Adapters/CollectionViewDataSource.swift]: Well, it's a single line comment 🤷🏼‍♀️

File: README.md [hunk 9]
```diff
@@ -220,3 +220,10 @@ Contributions are very welcome and highly appreciated! ❤️ Here's how to do i
 - If you have any questions feel free to create an issue with a `question` label;
 - If you have a feature request you can create an issue with a `Feature request` label;
 - If you found a bug feel free to create an issue with a `bug` label or open a PR with a fix.
+
+#### Image attributions
+
+[Coffee](https://pixabay.com/en/coffee-beans-coffee-beans-caffeine-3457587/)
+[Pomegranate fruit](https://pixabay.com/en/pomegranate-fruit-exotic-fruit-3383814/)
+[Cherries](https://pixabay.com/en/cherries-sweet-cherries-fruit-red-3433775/)
+[Strawberries](https://pixabay.com/en/strawberries-season-spring-fruit-3359755/)
```
Comment (daniloaliberti) [inline README.md]: @andersio Image attributions added here

File: Bento/Adapters/CollectionViewDataSource.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (sergdort) [reply Bento/Adapters/CollectionViewDataSource.swift]: Haha true 🙈

File: Bento/Adapters/CollectionViewDataSource.swift (hunk 1 above)
Comment (sergdort) [reply Bento/Adapters/CollectionViewDataSource.swift]: I believe it's a breaking change ¯\_(ツ)_/¯

**Comment by mluisbrown** [comment]
tenor-106031698

**Pull request merged**